pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -39,6 +39,24 @@ const (
 	ReasonConditionCheckFailed = "ConditionCheckFailed"
 )
 
+type SkippingReason string
+
+const (
+	WhenExpressionsSkip         SkippingReason = "WhenExpressionsSkip"
+	ConditionsSkip              SkippingReason = "ConditionsSkip"
+	ParentTasksSkip             SkippingReason = "ParentTasksSkip"
+	IsStoppingSkip              SkippingReason = "IsStoppingSkip"
+	IsGracefullyCancelledSkip   SkippingReason = "IsGracefullyCancelledSkip"
+	IsGracefullyStoppedSkip     SkippingReason = "IsGracefullyStoppedSkip"
+	MissingResultsReferenceSkip SkippingReason = "MissingResultsReferenceSkip"
+	None                        SkippingReason = "None"
+)
```

Comment:
I am trying to understand the list of reasons here and how they are utilized. 

These are great and I really appreciate the thorough explanation to the users why a task was skipped but at the same time, I think this might cause false positives or incorrect responses if communicated to the user.

How is reason ParentTasksSkip applicable with scoping the when expression to the task? No need to introduce a new reason for the current functionality which will be deprecated in 9 months.

For a task with multiple parents, there could be multiple reasons for skipping that task. Can we keep it simple for the first iteration and introduce these if they are visible in the status?

Reply:
had a conversation with @pritidesai about this, but following up here in case someone has the same questions

while there are multiple reasons for skipping, the precedence of the reasons is important and factored in here

https://github.com/tektoncd/pipeline/blob/1174247beb3c5146787899b8a5e2214c1d4867a1/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L193-L221

from pipeline state --> parent --> when expressions --> results references --> conditions (deprecated) 

either way, we need to know the exact reason a task was skipped so that we can handle the exception when when expressions are scoped to task in (specifically L#279):

https://github.com/tektoncd/pipeline/blob/1174247beb3c5146787899b8a5e2214c1d4867a1/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L266-L286